[WIP] Add large svd draft (#38)
dask/dask-blog: _posts/2019-07-02-very-large-svds.md

general, find a large number of sites in the genome sequence where
individuals differ from each other.  For example, humans have more
than 100 million variable sites in the genome, and modern studies
like the [UK BioBank](https://www.ukbiobank.ac.uk/) are working towards
sequencing the genomes of 1 million individuals or more.

In diploid species like humans, mice or mosquitoes, each individual
carries two genome sequences, one inherited from each parent.  At each
of the 100 million variable genome sites there will be two or more
"alleles" that a single genome might carry.  One way to think about
this is via the [Punnett
square](https://en.wikipedia.org/wiki/Punnett_square), which
represents the different possible genotypes that one individual might
carry at one of these variable sites:

<td>
<img src="https://upload.wikimedia.org/wikipedia/commons/9/93/Punnett_Square_Genetic_Carriers.PNG" alt="punnet square" height="40%" width="40%">
</td>

In the above there are three possible genotypes: AA, Aa, and aa.  For
computational genomics, these genotypes can be encoded as 0, 1, or 2.
In a study of a species with M genetic variants assayed in N

> individual samples, we can represent these genotypes as an (N x M)

Reviewer: ```suggestion
individual samples, we can represent these genotypes as an (M x N)
```

Sorry, got my dimensions the wrong way round.

```python
import dask.array as da

x = da.random.random(size=(1_000_000, 20_000), chunks=(20_000, 5_000))

u, s, v = da.linalg.svd_compressed(x, k=10, compute=True)
v.compute()
```

This call is no longer entirely lazy, and it recomputes `x` a couple times, but
it works, and it works using only a few GB of RAM on my consumer laptop.

It takes around 2min 30s time to compute that on my laptop.  That's great!  But we can do better

Adding GPUs (a 15 second SVD)
-----------------------------

We can increase performance considerably by using a multi-GPU machine.
Here is almost the same code, running in significantly less same time but we make the
following changes:

1.  We increase the size of the array by a factor of **10x**
2.  We switch out NumPy for CuPy, a GPU NumPy implementation
3.  We use a sixteen-GPU DGX-2 machine with NVLink interconnects between GPUs (this will dramatically decrease transfer time between workers)

On A DGX2 we can calculate an SVD on a 200GB Dask array between 10 and15 seconds: [SVD Multi-GPU Notebook](https://gist.github.com/quasiben/98ee254920837313946f621e103d41f4)

To see this run, we recommend viewing

> [the attached screencast](https://youtu.be/4X5yky2lvEw)

Author: My computer may be borked?  Can someone else check if this has audio?

---
layout: post
title: Very Large SVDs
tagline: Dask + CuPy + Zarr + Genomics

> author: Alistair Miles (Oxford Big Data Institute), Ben Zaitlen (NVIDIA), Matthew Rocklin (Coiled)

Reviewer: ```suggestion
author: Ben Zaitlen (NVIDIA), Matthew Rocklin (Coiled), Alistair Miles (Oxford University)
```

I don't deserve to be first author here :)

Reviewer 2: I think we should do it alphabetically by first name.  I believe that happens to coincide with the current ordering of names

Reviewer: That would feel a bit strange for me, my contribution to this nice work is very minor.

Reviewer 2: changed the author order

science, allows for exploring larger datasets and testing more
hypotheses in less time.  Practically, this means not simply a fast
SVD but an accelerated pipeline end-to-end, from data loading to
analysis, to understanding.

*We want to run an experiment in less time than it takes to make a cup of tea*

Performant SVDs w/ Dask
-----------------------

Now that we have that scientific background, let's transition back to talking about computation.

To stop Dask from holding onto the data we intentionally trigger computation as
we build up the graph.  This is a bit atypical in Dask calculations (we prefer
to have as much of the computation at once before computing) but useful given
the multiple-pass nature of this problem.  This was a fairly easy change, and
is available in [dask/dask #5041](https://github.com/dask/dask/pull/5041).

Additionally, we found that it was helpful to turn on moderately wide task
fusion.

```python
import dask
dask.config.set({"optimization.fuse.ave-width": 5})
```

Then things work fine
---------------------

We're going to try this SVD on a few different choices of hardware including:

1.  A MacBook Pro

> 2.  A DGX-2, an NVIDIA worksation with 16 high-end GPUs and fast network

Reviewer 2: ```suggestion
2.  A DGX-2, an NVIDIA worksation with 16 high-end GPUs and fast interconnect
```

3.  A twenty-node cluster on AWS

### Macbook Pro

We can happily perform an SVD on a 20GB array on a Macbook Pro

```python
import dask.array as da

x = da.random.random(size=(1_000_000, 20_000), chunks=(20_000, 5_000))

u, s, v = da.linalg.svd_compressed(x, k=10, compute=True)
v.compute()
```

This call is no longer entirely lazy, and it recomputes `x` a couple times, but
it works, and it works using only a few GB of RAM on a consumer laptop.

It takes around 2min 30s time to compute that on a laptop.
That's great!  It was super easy to try out, didn't require any special
hardware or setup, and in many cases is fast enough.
By working locally we can iterate quickly.

Now that things work, we can experiment with different hardware.

### Adding GPUs (a 15 second SVD)

*Disclaimer: one of the authors (Ben Zaitlen) works for NVIDIA*

We can dramatically increase performance by using a multi-GPU machine.
NVIDIA and other manufactures now make machines with multiple GPUs co-located in the same physical box.

> In the following section, we will run the calculations on a **DGX2**, a machine with 16 GPUs and fast network connection.

Reviewer 2: ```suggestion
In the following section, we will run the calculations on a **DGX2**, a machine with 16 GPUs and fast interconnect between the GPUs.
```
